File: tixspot/auth/endpoints.py
# pylint: disable=no-member


from connect_to_db import connect_to_db
from create_user import create_user
from connect_to_db import connect_to_db  # pylint: disable=import-error
from get_user import get_user
from fastapi import APIRouter
from datetime import datetime, timedelta
from typing import Union

from fastapi import Depends, Form, FastAPI, HTTPException, status, Header, Response, Request
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from jose import JWTError, jwt
from passlib.context import CryptContext
from pydantic import BaseModel
from typing_extensions import Annotated
from typing import Optional
from fastapi.middleware.cors import CORSMiddleware
from .auth import tokens_from_login, ACCESS_TOKEN_EXPIRE_MINUTES, REFRESH_TOKEN_EXPIRE_HOURS, new_tokens_using_refresh, validate_token
import json
router = APIRouter()


oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


class Token(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str


class User(BaseModel):
    first_name: Optional[str] = ''
    last_name: Optional[str] = ''
    username:Optional[str] = ''


class Login(BaseModel):
    email: str
    password: str

async def get_refresh_cookie(request):
    cookie_value = request.cookies.get("refresh_token")  
    
    if not cookie_value:
        return {"message": "Cookie not found"}
    print(refresh)
    tokens = await new_tokens_using_refresh(cookie_value)
    return(tokens)    

async def set_refresh_token(data, refresh_token):
    # Create a JSON response
    response = Response(content=json.dumps(data), media_type="application/json")
    # Set the first cookie
    response.set_cookie(
        key="refresh_token",
        value=refresh_token,
        samesite='none',
        secure=True
    )
    return(response)




# creates new access and refresh tokens, need to send username and password in formdata


@router.post("/login", response_model=Token)
async def login_for_access_token(
    email: Annotated[str, Form()], password: Annotated[str, Form()]
):
    (access_token, refresh_token) = await tokens_from_login(
        email, password, ACCESS_TOKEN_EXPIRE_MINUTES, REFRESH_TOKEN_EXPIRE_HOURS)
    data = {
        "access_token": access_token,
       "token_type": "bearer"
    }
    
    # Create a JSON response
    response = await set_refresh_token(data,refresh_token)
    return response



@router.post("/register", response_model=Token)
async def register(email: str = Form(),password:str=Form()):
    client = connect_to_db()

    user_id = create_user(client['tixspot'],email,password)
    if not user_id:
        raise HTTPException(status_code=409, detail="User already exists")
    (access_token, refresh_token) = await tokens_from_login(
        email, password, ACCESS_TOKEN_EXPIRE_MINUTES, REFRESH_TOKEN_EXPIRE_HOURS)
    data={"access_token": access_token, "token_type": "bearer"}
    print(data)
    response = await set_refresh_token(data,refresh_token)
    return(response)


@router.get("/refresh")  # creates new tokens using a refresh token
async def refresh(
    request: Request,refresh: Annotated[Union[str, None], Header()] = None):
    tokens = await get_refresh_cookie(request)
    refresh_token=tokens.pop("refresh_token")
    print(refresh_token)
    return {"tokens": tokens}


# @router.get("/protected")  # protected routes can be used this way
# async def protected(
#     authorization: Annotated[Union[str, None], Depends(oauth2_scheme)] = None
# ):
#     user = await validate_token(authorization, 'access')
#     print(user['username'])
#     return [{"user": user['username']}]

File: tixspot/auth/endpoints.go
package main

import (
	"encoding/json"
	"net/http"
	"github.com/gin-gonic/gin"
	"github.com/dgrijalva/jwt-go"
	"time"
	"errors"
)

var (
	// Define a secret key for signing tokens
	jwtKey = []byte("your_secret_key")
)

// Token represents the structure of the authentication token
type Token struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	TokenType    string `json:"token_type"`
}

// User represents a user in the system
type User struct {
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Username  string `json:"username"`
}

// Login represents the login request body
type Login struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

// Function to create a user (implement database logic here)
func createUser(email, password string) (string, error) {
	// Check if user exists and create new user logic here
	return "user_id", nil // return user ID or an error
}

// Function to generate tokens (implement logic here)
func tokensFromLogin(email, password string) (string, string, error) {
	// Validate user credentials and generate tokens
	accessToken := "generated_access_token"
	refreshToken := "generated_refresh_token"
	return accessToken, refreshToken, nil
}

// Middleware to check for token validity
func validateToken(tokenStr string) (map[string]interface{}, error) {
	token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
		// Check token signing method
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("invalid signing method")
		}
		return jwtKey, nil
	})

	if err != nil {
		return nil, err
	}

	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
		return claims, nil
	}
	return nil, errors.New("invalid token")
}

// Set refresh cookie
func setRefreshCookie(c *gin.Context, refreshToken string) {
	c.SetCookie("refresh_token", refreshToken, 3600*24*30, "/", "", true, true) // 30 days
}

func main() {
	router := gin.Default()

	router.POST("/login", func(c *gin.Context) {
		var login Login
		if err := c.ShouldBindJSON(&login); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
			return
		}

		accessToken, refreshToken, err := tokensFromLogin(login.Email, login.Password)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "invalid credentials"})
			return
		}

		data := Token{
			AccessToken:  accessToken,
			TokenType:    "bearer",
			RefreshToken: refreshToken,
		}
		setRefreshCookie(c, refreshToken)
		c.JSON(http.StatusOK, data)
	})

	router.POST("/register", func(c *gin.Context) {
		var login Login
		if err := c.ShouldBindJSON(&login); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
			return
		}

		userID, err := createUser(login.Email, login.Password)
		if err != nil {
			c.JSON(http.StatusConflict, gin.H{"error": "user already exists"})
			return
		}

		accessToken, refreshToken, err := tokensFromLogin(login.Email, login.Password)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "invalid credentials"})
			return
		}

		data := Token{
			AccessToken:  accessToken,
			TokenType:    "bearer",
			RefreshToken: refreshToken,
		}
		setRefreshCookie(c, refreshToken)
		c.JSON(http.StatusOK, data)
	})

	router.GET("/refresh", func(c *gin.Context) {
		refreshToken, err := c.Cookie("refresh_token")
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "cookie not found"})
			return
		}

		// Here you would validate the refresh token and issue new tokens
		// For simplicity, just returning a dummy access token
		newAccessToken := "new_generated_access_token"
		c.JSON(http.StatusOK, gin.H{"access_token": newAccessToken})
	})

	router.Run(":8080")
}
